### PR TITLE
Util/json glaze

### PR DESCRIPTION
**Pull request by AlemSnyder**
Convert JSON parsing to glaze. Create structs to read json data into.

### CODE REVIEW

**Comment by AlemSnyder** [comment]
This isn't ready for review because I need to merge day-night-cycle-controls (#75) because I need to add the json reading of tile entities.

**Comment by AlemSnyder** [comment]
I'm marking this ready for review because I don't know how to fix the issues.

File: src/world/entity/model.hpp [hunk 1]
```diff
@@ -4,19 +4,52 @@
...
+struct remapping_t {
+    std::map<ColorInt, ColorInt> map;
+
+    void
+    read(std::string input);
+    
+    [[nodiscard]] std::string
+    write() const;
+
+};
```
Comment (egelja) [inline src/world/entity/model.hpp:37]: https://godbolt.org/z/KEjn4n7nh

File: src/world/entity/object_handler.cpp [hunk 2]
```diff
@@ -9,31 +8,39 @@ namespace world {
...
         );
+
+        auto ec = glz::read_json(object_data, content);
+        if (ec) {
+            LOG_ERROR(logging::file_io_logger, "{}", glz::format_error(ec, content));
+            return;
+        }
+    } else {
         return;
     }
```
Comment (egelja) [inline src/world/entity/object_handler.cpp:38]: use `read_file_json` like the comment above

File: src/world/terrain/generation/land_generator.cpp [hunk 3]
```diff
@@ -32,36 +31,24 @@ namespace terrain {
...
                 stamp_generator = std::make_shared<stamps::FromGrid>(region);
                 break;
             [[unlikely]] default:
-                LOG_WARNING(
-                    logging::terrain_logger,
-                    "Terrain stamp type {} is not valid. Must be one of Position, "
-                    "Radius, Grid",
-                    type
-                );
+                LOG_WARNING(logging::terrain_logger, "-\\:)/-");
```
Comment (egelja) [inline src/world/terrain/generation/land_generator.cpp:51]: come on use a good error message

File: src/world/terrain/generation/land_generator.cpp [hunk 4]
```diff
@@ -143,36 +130,177 @@ JsonToTile::get_volume(glm::imat2x2 center, std::default_random_engine& rand_eng
...
+                    glz::format_error(ec_3, material_data.material.str)
+                );
+            } else {
+                goto end_parse_material;
+            }
+
+            LOG_CRITICAL(logging::file_io_logger, "could not parse material");
+        }
+
+end_parse_material:
```
Comment (egelja) [inline src/world/terrain/generation/land_generator.cpp:229]: Use a `std::variant`

File: src/world/terrain/generation/land_generator.cpp [hunk 5]
```diff
@@ -143,36 +130,177 @@ JsonToTile::get_volume(glm::imat2x2 center, std::default_random_engine& rand_eng
...
+                        materials_w_color.insert(
+                            {material_id, std::set<ColorId>(
+                                              color_data_v.value().begin(),
+                                              color_data_v.value().end()
+                                          )}
+
+                        );
+                    }
+                }
             }
```
Comment (egelja) [inline src/world/terrain/generation/land_generator.cpp:300]: what is this code

File: src/world/terrain/generation/terrain_genreration_types.hpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+enum class generation_stamp_type {
+    GRID,
+    RADIUS,
+    POSITION,
+};
+
+struct material_designation_t {
+    glz::raw_json material; // bool or int
+    glz::raw_json color;    // bool or int
```
Comment (egelja) [inline src/world/terrain/generation/terrain_genreration_types.hpp:25]: https://en.cppreference.com/w/cpp/utility/variant

File: src/world/terrain/generation/terrain_genreration_types.hpp [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    assert(false && "Not Implemented");
+    return {};
+}
+
+template <>
+inline glz::detail::any_t::operator terrain::generation::stamp_generation_radius_data_t(
+) const {
+    assert(false && "Not Implemented");
+    return {};
+}
```
Comment (egelja) [inline src/world/terrain/generation/terrain_genreration_types.hpp:146]: i bet this is related to you using `raw_json` and will be fixed once you use a variant

File: src/world/terrain/generation/terrain_genreration_types.hpp [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    Dim stop;
+    Dim data;
+
+    layer_effect_add add_directions;
+
+    bool
+    operator<(const layer_effect_data_t& other) const {
+        return start < other.start;
+    }
+};
```
Comment (egelja) [inline src/world/terrain/generation/terrain_genreration_types.hpp:85]: do the spaceship operator instead

File: src/world/terrain/material.hpp [hunk 9]
```diff
@@ -49,22 +53,73 @@ namespace terrain {
...
     // int8_t deterioration from water
 };
 
+struct material_color_t {
+    std::string color_name;
+    ColorInt hex_color;
+
+    void
+    read_hex_color(std::string hex_string) {
+        hex_color = std::stoll(hex_string, 0, 16);
```
Comment (egelja) [inline src/world/terrain/material.hpp:78]:
```suggestion
        hex_color = std::stoll(hex_string, nullptr, 16);
```

ALEM STOP USING 0 FOR nullptr

File: src/world/terrain/material.hpp [hunk 10]
```diff
@@ -49,22 +53,73 @@ namespace terrain {
...
+    }
+
+    std::string
+    string_hex_color() const {
+        std::stringstream stream;
+        stream << std::hex << hex_color;
+        std::string result(stream.str());
+        return result;
+    }
```
Comment (egelja) [inline src/world/terrain/material.hpp:87]:
```suggestion
    std::string
    write_hex_color() const {
        return fmt::format("{:08X}", hex_color);
    }
```

File: src/world/terrain/material.hpp [hunk 11]
```diff
@@ -49,22 +53,73 @@ namespace terrain {
...
+
+    void
+    read_color_name(std::string color_name_) {
+        color_name = color_name_;
+    }
+
+    std::string
+    write_color_name() const {
+        return color_name;
+    }
```
Comment (egelja) [inline src/world/terrain/material.hpp:97]: you don't need custom functions if they just do what the default would

File: src/world/terrain/material.hpp [hunk 12]
```diff
@@ -239,3 +294,29 @@ material_in(
...
+
+template <>
+struct glz::meta<terrain::material_color_t> {
+    // clang-format off
+    static constexpr auto value =
+        object("hex_color",  custom<&terrain::material_color_t::read_hex_color,
+                                    &terrain::material_color_t::string_hex_color>,
+               "color_name", custom<&terrain::material_color_t::color_name,
+                                    &terrain::material_color_t::color_name>);
+    // clang-format on
```
Comment (egelja) [inline src/world/terrain/material.hpp:315]:
```suggestion
    using T = terrain::material_color_t;
    // clang-format off
    static constexpr auto value = object(
        "hex_color",  custom<&T::read_hex_color, &T::write_hex_color>,
        "color_name", &T::color_name,
    );
    // clang-format on
```

File: src/world/terrain/material.hpp [hunk 13]
```diff
@@ -239,3 +294,29 @@ material_in(
...
+};
+
+template <>
+inline glz::detail::any_t::operator terrain::grass_data_t() const {
+    assert(false && "Not Implemented");
+    return {};
+}
```
Comment (egelja) [inline src/world/terrain/material.hpp:322]: stop using `raw_json`

File: src/world/terrain/material.hpp [hunk 14]
```diff
@@ -49,22 +53,73 @@ namespace terrain {
...
+
+    std::optional<grass_data_t> gradient;
+};
+
+struct all_materials_reader_t {
+    std::map<glz::sv, glz::raw_json> data;
+};
+
+struct all_materials_t {
+    std::map<std::string, material_t> data;
```
Comment (egelja) [inline src/world/terrain/material.hpp:120]: You don't need two different ones
https://godbolt.org/z/KEjn4n7nh

**Review by egelja — CHANGES_REQUESTED** [review]

File: src/world/entity/model.hpp [hunk 15]
```diff
@@ -4,19 +4,74 @@
...
+        for (const auto& [key, value] : map) {
+            std::stringstream stream_key;
+            stream_key << std::hex << key;
+            std::string str_key(stream_key.str());
+
+            std::stringstream stream_value;
+            stream_value << std::hex << value;
+            std::string str_value(stream_value.str());
+
+            res.insert({str_key, str_value});
```
Comment (egelja) [inline src/world/entity/model.hpp:52]: change to use fmt like in the godbolt example

iostream is about 20-30x slower

File: src/world/entity/model.hpp [hunk 16]
```diff
@@ -213,3 +269,16 @@ class ModelController : virtual public gui::gpu_data::GPUDataElementsInstanced {
...
+
+template <>
+struct glz::meta<world::entity::remapping_t> {
+
+    // clang-format off
+    static constexpr auto value =
+        object("map",  custom<&world::entity::remapping_t::read_map,
+                              &world::entity::remapping_t::write_map>);
+    // clang-format on
+};
```
Comment (egelja) [inline src/world/entity/model.hpp:284]:
```suggestion
struct glz::meta<world::entity::remapping_t> {
    using T = world::entity::remapping_t
    // clang-format off
    static constexpr auto value = object(
        "map",  custom<&T::read_map, &T::write_map>
    );
    // clang-format on
};
```

File: src/world/entity/object_handler.cpp [hunk 17]
```diff
@@ -9,31 +8,37 @@ namespace world {
...
+
+    object_t object_data;
 
     // read contents from path
-    auto contents = files::open_data_file(object_path);
+    auto contents = files::open_data_file(descriptor.path);
     if (contents.has_value()) {
-        contents.value() >> object_json;
+        std::string content;
+        *contents.value() >> content;
```
Comment (egelja) [inline src/world/entity/object_handler.cpp:27]:
```suggestion
        *contents >> content;
```

Either use `*` or `.value()`, but not both

**Review by egelja — CHANGES_REQUESTED** [review]

File: src/world/terrain/generation/land_generator.cpp (hunk 5 above)
Comment (AlemSnyder) [reply src/world/terrain/generation/land_generator.cpp:300]: hehe :3

**Comment by AlemSnyder** [comment]
I might not change the file reading mechanisms, and the unordered things will be done in a different pr. Everything else is good and helpful.

File: src/world/terrain/generation/land_generator.cpp (hunk 4 above)
Comment (AlemSnyder) [reply src/world/terrain/generation/land_generator.cpp:229]: can't put a vetor in a varient

Comment (egelja) [reply src/world/terrain/generation/land_generator.cpp]: https://godbolt.org/z/77a849x6e

Comment (AlemSnyder) [reply src/world/terrain/generation/land_generator.cpp]: Ya I realized i was wrong. I was going to do this then realized the next part is... bad...

The spec says no arrays, I remember that as vectors for some reason.

Comment (egelja) [reply src/world/terrain/generation/land_generator.cpp]: Yea arrays makes sense

File: src/world/terrain/generation/terrain_genreration_types.hpp (hunk 7 above)
Comment (AlemSnyder) [reply src/world/terrain/generation/terrain_genreration_types.hpp:167]: its not

File: src/util/loading.cpp [hunk 18]
```diff
@@ -16,62 +16,110 @@ void
...
+        int num_biomes =
+            manifest.biomes.has_value() ? manifest.biomes.value().size() : 0;
+        int num_entities =
+            manifest.entities.has_value() ? manifest.entities.value().size() : 0;
+
+        LOG_DEBUG(
+            logging::file_io_logger, "Loading manifest {} with {} biomes, {} entities.",
+            manifest.name, num_biomes, num_entities
+        );
+        if (manifest.entities.has_value()) {
```
Comment (egelja) [inline src/util/loading.cpp:39]:
```suggestion
        if (manifest.entities) {
```

File: src/util/loading.cpp [hunk 19]
```diff
@@ -16,62 +16,110 @@ void
...
+            // iterate through objects in manifest and queue them to be loaded
+            for (const manifest::descriptor_t& entity_data :
+                 manifest.entities.value()) {
```
Comment (egelja) [inline src/util/loading.cpp:42]:
```suggestion
            for (const manifest::descriptor_t& entity_data : *manifest.entities) {
```

File: src/util/loading.cpp [hunk 20]
```diff
@@ -16,62 +16,110 @@ void
...
+}
 
-            Json::CharReaderBuilder builder;
-            builder["collectComments"] = false;
+// I know this entire things is copied. I know it's bad.
+// It is like this because I don't know if I can
+//  1 use opengl without a screen (GitHub)
+//  2 conveniently get errors back
+int
+load_manifest_test() {
```
Comment (egelja) [inline src/util/loading.cpp:61]: update this with comments from above

File: src/world/biome.cpp [hunk 21]
```diff
@@ -48,22 +48,31 @@ GrassData::GrassData(const Json::Value& json_grass_data) {
     grass_grad_length_ = grass_colors_.size();
 }
 
-Biome::Biome(const biome_json_data& biome_data, size_t seed) :
+GrassData::GrassData(const std::optional<grass_data_t>& grass_data) :
+    GrassData(grass_data.value_or(grass_data_t())) {
+    if (!grass_data.has_value()) {
```
Comment (egelja) [inline src/world/biome.cpp:53]:
```suggestion
    if (!grass_data) {
```

File: src/world/biome.cpp [hunk 22]
```diff
@@ -398,20 +379,27 @@ Biome::get_colors_inverse_map() const {
...
 
 biome_json_data
-Biome::get_json_data_(const std::string& biome_name) {
-    std::filesystem::path biome_json_path = files::get_data_path() / biome_name;
+Biome::get_json_data(const std::filesystem::path& biome_folder_path) {
+    auto biome_data = files::read_json_from_file<terrain::generation::biome_data_t>(
+        biome_folder_path / "biome_data.json"
+    );
+
+    if (!biome_data.has_value()) {
```
Comment (egelja) [inline src/world/biome.cpp:387]:
```suggestion
    if (!biome_data) {
```

File: src/world/biome.cpp [hunk 23]
```diff
@@ -398,20 +379,27 @@ Biome::get_colors_inverse_map() const {
...
-        biome_file.value() >> biome_data;
+    auto materials = files::read_json_from_file<all_materials_t>(
+        biome_folder_path / "materials.json"
+    );
 
-    Json::Value materials_data;
-    auto materials_file = files::open_data_file(biome_json_path / "materials.json");
-    if (materials_file.has_value())
-        materials_file.value() >> materials_data;
+    if (!materials.has_value()) {
```
Comment (egelja) [inline src/world/biome.cpp:395]:
```suggestion
    if (!materials) {
```

File: src/world/biome.cpp [hunk 24]
```diff
@@ -398,20 +379,27 @@ Biome::get_colors_inverse_map() const {
...
+        return {};
+    }
 
-    return {biome_name, std::move(biome_data), std::move(materials_data)};
+    terrain::generation::biome_json_data data(
+        biome_data.value().name, biome_data.value(), materials.value()
```
Comment (egelja) [inline src/world/biome.cpp:400]:
```suggestion
        biome_data->name, *biome_data, *materials
```

File: src/world/entity/object_handler.cpp [hunk 25]
```diff
@@ -9,31 +9,29 @@ namespace world {
...
     // read contents from path
-    auto contents = files::open_data_file(object_path);
-    if (contents.has_value()) {
-        contents.value() >> object_json;
-    } else {
-        LOG_WARNING(
-            logging::file_io_logger, "Cannot open file {}.", object_path.string()
+    auto object_data = files::read_json_from_file<object_t>(descriptor.path);
+
+    if (!object_data.has_value()) {
```
Comment (egelja) [inline src/world/entity/object_handler.cpp:23]:
```suggestion
    if (!object_data) {
```

File: src/world/entity/object_handler.cpp [hunk 26]
```diff
@@ -47,7 +45,7 @@ ObjectHandler::read_object(std::filesystem::path object_path) {
     // on the main thread
     ided_objects.emplace(
         std::piecewise_construct, std::forward_as_tuple(std::move(identification)),
-        std::forward_as_tuple(object_json, object_path)
+        std::forward_as_tuple(object_data.value(), descriptor)
```
Comment (egelja) [inline src/world/entity/object_handler.cpp:48]:
```suggestion
        std::forward_as_tuple(*object_data, descriptor)
```

File: src/world/terrain/generation/land_generator.hpp [hunk 27]
```diff
@@ -198,11 +205,10 @@ class LandGenerator {
     size_t current_region;
     size_t current_sub_region;
 
-    // const std::map<MaterialId, const Material>& materials;
-    // const std::set<Material*> elements_can_stamp_;
-    // const Material* material_;
+    // const std::map<MaterialId, const material_t>& materials;
+    // const std::set<material_t*> elements_can_stamp_;
+    // const material_t* material_;
```
Comment (egelja) [inline src/world/terrain/generation/land_generator.hpp:210]: what

File: src/world/terrain/generation/land_generator.hpp [hunk 28]
```diff
@@ -258,41 +264,20 @@ class LandGenerator {
...
-
 class AddToTop {
     const MaterialGroup elements_above_;
 
     const MaterialGroup elements_can_overwrite_;
     const MaterialId stamp_material_id_; // MaterialId that will be set
     const ColorId stamp_color_id_;       // ColorId that will be set
 
-    std::set<AddMethod> data_;
+    std::set<layer_effect_data_t> data_;
```
Comment (egelja) [inline src/world/terrain/generation/land_generator.hpp:274]: unordered

File: src/world/terrain/generation/terrain_genreration_types.hpp [hunk 29]
```diff
@@ -0,0 +1,167 @@
...
+            return stop_c;
+        }
+        auto start_c = start <=> other.start;
+        if (start_c != std::strong_ordering::equivalent) {
+            return start_c;
+        }
+
+        // we want to compare directions like this because going to a position is
+        // bigger than adding some value assuming the stop is the same
+        auto directions_c = (int)add_directions <=> (int)other.add_directions;
```
Comment (egelja) [inline src/world/terrain/generation/terrain_genreration_types.hpp:97]: static cast

File: src/world/terrain/material.cpp [hunk 30]
```diff
@@ -59,4 +57,110 @@ TerrainColorMapping::assign_color_mapping(
...
 }
 
+MaterialGroup::MaterialGroup(const std::vector<generation::material_designation_t>& data
+) :
+    contain_all_materials(false) {
+    // want to return a group that represents the given data
+    // There will be elements with no requirements on the color
+    //    std::set<MaterialId> materials;
+    // amd some with requirements on the color
+    //    std::map<MaterialId, std::set<ColorId>> materials_w_color;
```
Comment (egelja) [inline src/world/terrain/material.cpp:67]: unordered

File: src/world/terrain/material.cpp [hunk 31]
```diff
@@ -59,4 +57,110 @@ TerrainColorMapping::assign_color_mapping(
...
+        // read the material id from the data
+        // MaterialId mat_id = material_data.material;
+
+        const auto& material = material_data.material;
+        const auto& color = material_data.color;
+
+        std::unordered_set<ColorId> colors;
+        std::unordered_set<MaterialId> materials;
+
+        {
```
Comment (egelja) [inline src/world/terrain/material.cpp:79]: this should be extracted into a function

File: src/world/terrain/material.cpp [hunk 32]
```diff
@@ -59,4 +57,110 @@ TerrainColorMapping::assign_color_mapping(
...
+            switch (material.index()) {
```
Comment (egelja) [inline src/world/terrain/material.cpp:80]: NO NO NO NO NO NO NO NO NO this is UB

Use a visistor: https://en.cppreference.com/w/cpp/utility/variant/visit

File: src/world/terrain/material.cpp [hunk 33]
```diff
@@ -59,4 +57,110 @@ TerrainColorMapping::assign_color_mapping(
...
+
+                        break;
+                    }
+
+                default:
+                    break;
+            }
+        }
+
+        {
```
Comment (egelja) [inline src/world/terrain/material.cpp:117]: this should be extracted into a function

File: src/world/terrain/material.cpp [hunk 34]
```diff
@@ -59,4 +57,110 @@ TerrainColorMapping::assign_color_mapping(
...
+            switch (color.index()) {
```
Comment (egelja) [inline src/world/terrain/material.cpp:118]: again, use a visitor

https://en.cppreference.com/w/cpp/utility/variant/visit

File: src/world/terrain/material.cpp [hunk 35]
```diff
@@ -59,4 +57,110 @@ TerrainColorMapping::assign_color_mapping(
...
+            switch (material.index()) {
+                case 0:
+                    { // bool
+
+                        bool all_materials = std::get<bool>(material);
+                        if (all_materials) {
+                            set_all();
+                            return;
+                        } else {
+                            // wut? why would you do this?
```
Comment (egelja) [inline src/world/terrain/material.cpp:89]: log a mesage then

File: src/world/terrain/material.cpp [hunk 36]
```diff
@@ -59,4 +57,110 @@ TerrainColorMapping::assign_color_mapping(
...
+
+        {
+            switch (material.index()) {
+                case 0:
+                    { // bool
+
+                        bool all_materials = std::get<bool>(material);
+                        if (all_materials) {
+                            set_all();
+                            return;
```
Comment (egelja) [inline src/world/terrain/material.cpp:87]: this return bypasses the colors thing below too --- is that desired?

**Review by egelja — CHANGES_REQUESTED** [review]

File: src/util/loading.cpp (hunk 20 above)
Comment (AlemSnyder) [reply src/util/loading.cpp:61]: Sorry this wasn't clear in the comments. The difference between `load_manifest` and `load_manifest_test` is that load_manifest_test does not use opengl, and each "task" is run in series and if any fail the function returns.

Comment (egelja) [reply src/util/loading.cpp:61]: No i mean update this function with the comments i made on `load_manifest` lol

File: src/world/terrain/generation/land_generator.hpp (hunk 27 above)
Comment (AlemSnyder) [reply src/world/terrain/generation/land_generator.hpp:210]: hm must be old

File: src/world/terrain/generation/land_generator.hpp (hunk 28 above)
Comment (AlemSnyder) [reply src/world/terrain/generation/land_generator.hpp:274]: 1) wrong pr
2) I used the ordering somewhere. I think I will again to determine the color a tile should be.

File: src/world/terrain/material.cpp (hunk 36 above)
Comment (AlemSnyder) [reply src/world/terrain/material.cpp:87]: yes. `set_all` clears the mat and set stored in the class.

File: src/util/files.hpp [hunk 37]
```diff
@@ -80,22 +93,50 @@ open_resource_file(std::filesystem::path path) {
...
+        LOG_ERROR(logger, "Path {} does not exist.", path);
+        return {};
+    }
+
+    T data;
+    auto ec = glz::read_file_json(data, path.c_str(), std::string{});
+
+    if (ec) {
+        LOG_ERROR(
+            logger, "Failed to read {} from path {}. Error: {}", typeid(T).name(), path,
```
[the remark this answers is not in the document]
Comment (AlemSnyder) [reply src/util/files.hpp:134]: Would be a headake to implement though

File: src/util/loading.cpp [hunk 38]
```diff
@@ -18,58 +18,102 @@ load_manifest() {
...
-        if (!manifest.isObject()) [[unlikely]] {
-            LOG_WARNING(
-                logging::file_io_logger,
-                "Manifest file {} did not produce a valid JSON object.", manifest_file
-            );
-            continue;
+        manifest::manifest_t& manifest = *manifest_opt;
+
+        int num_biomes = manifest.biomes ? (*manifest.biomes).size() : 0;
+        int num_entities = manifest.entities ? (*manifest.entities).size() : 0;
```
Comment (AlemSnyder) [inline src/util/loading.cpp:31]:
```suggestion
        int num_biomes = manifest.biomes ? manifest.biomes->size() : 0;
        int num_entities = manifest.entities ? manifest.entities->size() : 0;
```

File: src/world/terrain/generation/terrain_genreration_types.hpp [hunk 39]
```diff
@@ -0,0 +1,168 @@
...
+    std::vector<std::pair<int, int>> positions;
+};
+
+struct generation_stamp_t {
+    generation_stamp_type type;
+    MaterialId material_id;
+    ColorId color_id;
+    std::vector<material_designation_t> can_override;
+    int size;
+    int DS;
```
Comment (AlemSnyder) [inline src/world/terrain/generation/terrain_genreration_types.hpp:50]: no abbr

File: src/world/terrain/generation/terrain_genreration_types.hpp [hunk 40]
```diff
@@ -0,0 +1,168 @@
...
+    int height;
+    int DH;
```
Comment (AlemSnyder) [inline src/world/terrain/generation/terrain_genreration_types.hpp:52]: no abbr

File: src/world/terrain/generation/terrain_genreration_types.hpp [hunk 41]
```diff
@@ -0,0 +1,168 @@
...
+    std::optional<stamp_generation_grid_data_t> grid;
+    std::optional<stamp_generation_position_data_t> position;
+    std::optional<stamp_generation_radius_data_t> radius;
+
+    int DC;
```
Comment (AlemSnyder) [inline src/world/terrain/generation/terrain_genreration_types.hpp:57]: no abbr

**Pull request merged**